common/linkutils/link.go

```diff
@@ -0,0 +1,109 @@
+package linkutils
+
+import (
+	"errors"
+	"net/url"
+	"strings"
+)
+
+var ErrInvalidURL = errors.New("URL format is incorrect")
+
+// URL focuses on query params and treats the rest as immutable string.
+// You can use any URL string to construct this object.
+// Its primary goal is to "Expose Query Manipulation".
+// Under the hood it uses url.URL library.
+type URL struct {
```

Comment:
This is good. I like the basic idea here. However the specific implementation is problematic - reason being that URLs have fairly esoteric escape logic. If memory serves, there are like 6 distinct parts of a URL, and each part has its own escape rules. Trying to roll your own is going to fall short (not that you can't do it, just that you'll eventually end up re-inventing the wheel).

One idea is to look at something like [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570). There's a good library which already implements this: https://github.com/jtacoma/uritemplates . It handles the escape logic nicely. You can still have this wrapper layer to simplify things, but I think string concatenation alone will eventually cause problems.

Another idea is to take these URL fields and use it to construct (on the fly) a `url.URL` instance (with the scheme, host, path, query params, fragment, etc all set explicitly). Then you can rely on the URL to properly escape things.